### PR TITLE
Spir-V pipeline improvements

### LINKED ISSUE #5674: [Vulkan] VK_ERROR_INITIALIZATION_FAILED
**Describe the bug (REQUIRED)**
Probably my fault somewhere on fragment shader but it is not possible to understand from this error message.

```cpp
[mvk-error] VK_ERROR_INITIALIZATION_FAILED: Render pipeline compile failed (Error code 1):
Link failed: fragment input user(locn1) was not found in vertex shader outputs.
ERROR:GRAPHICS: Vulkan Error (../src/vulkan/graphics_vulkan.cpp:1336) VK_ERROR_INITIALIZATION_FAILED
Assertion failed: (0), function GetOrCreatePipeline, file ../src/vulkan/graphics_vulkan.cpp, line 1336.
```

**Defold version (REQUIRED):**

Screen Shot 2021-03-21 at 20 57 37

**Platforms (REQUIRED):**
 - Platforms: macOS
 - OS: 10.15.7 

**Logs (OPTIONAL):**
[_crash -vulkan.zip](https://github.com/defold/defold/files/6178029/_crash.-vulkan.zip)

### PR DESCRIPTION
**Pull request by Jhonnyg**
Improved the SPIR-V shader pipeline:
* Better error feedback when shader inputs and outputs mismatch during the build process
* Removed redundancy when producing shader resource bindings since we now run optimisation passes on the SPIR-V source
* Output name hashes for shader bindings instead of doing it in runtime

Fixes #5674

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: engine/gamesys/src/gamesys/resources/res_material.cpp [hunk 1]
```diff
@@ -210,6 +210,12 @@ namespace dmGameSystem
         if (r == dmResource::RESULT_OK)
         {
             dmRender::HMaterial material = dmRender::NewMaterial(render_context, resources.m_VertexProgram, resources.m_FragmentProgram);
+            if (!material)
+            {
+                dmResource::Release(params.m_Factory, (void*)resources.m_VertexProgram);
+                dmResource::Release(params.m_Factory, (void*)resources.m_FragmentProgram);
+                return dmResource::RESULT_DDF_ERROR;
+            }
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/resources/res_material.cpp:218]: Avoids a hard crash when the program can't be created

File: engine/graphics/proto/graphics/graphics_ddf.proto [hunk 2]
```diff
@@ -196,19 +196,21 @@ message ShaderDesc
...
     message Shader
     {
         required Language        language              = 1;
         optional bytes           source                = 2;
         repeated ResourceBinding uniforms              = 4;
-        repeated ResourceBinding attributes            = 5;
-        optional bool            variant_texture_array = 6 [default = false];
+        repeated ResourceBinding inputs                = 5;
+        repeated ResourceBinding outputs               = 6;
+        optional bool            variant_texture_array = 7 [default = false];
```
Comment (Jhonnyg) [inline engine/graphics/proto/graphics/graphics_ddf.proto:213]: Changed the order a bit here, but it's build/engine only so should be fine. Editor uses bob for building so no changes should be needed for editor either

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 3]
```diff
@@ -2135,23 +2131,35 @@ namespace dmGraphics
...
+            shader->m_OutputCount = ddf->m_Outputs.m_Count;
+
+            for (uint32_t i=0; i < ddf->m_Outputs.m_Count; i++)
+            {
+                ShaderResourceBinding& res = shader->m_Outputs[i];
+                res.m_Binding              = ddf->m_Outputs[i].m_Binding;
+                res.m_Set                  = ddf->m_Outputs[i].m_Set;
+                res.m_Type                 = ddf->m_Outputs[i].m_Type;
+                res.m_NameHash             = ddf->m_Outputs[i].m_NameHash;
+                res.m_Name                 = strdup(ddf->m_Outputs[i].m_Name);
```
Comment (JCash) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:2162]: Not sure if it's worth it but these two segments could be put into a function. it's an edge case I think.

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 4]
```diff
@@ -2220,8 +2228,28 @@ namespace dmGraphics
...
+                    input_found = true;
+                    break;
+                }
+            }
+
+            if (!input_found)
+            {
+                dmLogError("Input of fragment shader '%s' not written by vertex shader", fragment_module->m_Inputs[i].m_Name);
+                return false;
+            }
```
Comment (JCash) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:2250]: I feel like this check should be in the shader build pipeline?

At runtime we don't change the vertex/fragment programs, only the whole material. So this situation shouldn't come up at runtime?

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 5]
```diff
@@ -2250,11 +2278,11 @@ namespace dmGraphics
         HashState64 program_hash;
         dmHashInit64(&program_hash, false);
 
-        if (vertex_module->m_AttributeCount > 0)
+        if (vertex_module->m_InputCount > 0)
```
Comment (JCash) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:2281]: This if-statement is redundant. The for-loop does the job.

**Review by JCash — DISMISSED** [review]
Looks good.
Although if possible, I'd like to move the input/output check to the pipeline.

**Pull request merged**